File: openFaces/source/META-INF/resources/openfaces/timetable/weekSwitcher.js
/*
 * OpenFaces - JSF Component Library 2.0
 * Copyright (C) 2007-2011, TeamDev Ltd.
 * licensing@openfaces.org
 * Unless agreed in writing the contents of this file are subject to
 * the GNU Lesser General Public License Version 2.1 (the "LGPL" License).
 * This library is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * Please visit http://openfaces.org/licensing/ for more details.
 */

O$.WeekSwitcher = {
  _init: function(switcherId,
                  timeTableId,
                  firstDay,
                  pattern,
                  locale,
                  stylingParams,
                  enabled,
                  splitter) {
    var dtf = O$.getDateTimeFormatObject(locale);
    var switcher = O$.initComponent(switcherId, {rollover: stylingParams.rolloverClass}, {
      _firstDay: dtf.parse(firstDay, "dd/MM/yyyy"),
      _timeTableId: timeTableId,
      _pattern: pattern,
      _locale: locale,
      _splitter: splitter
    });

    var textId = switcherId + "::text";
    switcher._text = O$(textId);
    if (switcher._text) {
      O$.initComponent(textId, {rollover: stylingParams.textRolloverClass});
    }

    if (enabled) {
      var previousButtonId = switcherId + "::previous_button";
      O$.initComponent(previousButtonId, {rollover: stylingParams.previousButtonRolloverClass});
      switcher._previousButton = O$(previousButtonId);
      var previousButton = switcher._previousButton;

      var nextButtonId = switcherId + "::next_button";
      O$.initComponent(nextButtonId, {rollover: stylingParams.nextButtonRolloverClass});
      switcher._nextButton = O$(nextButtonId);
      var nextButton = switcher._nextButton;

      previousButton.onmouseup = function () {
        O$.setStyleMappings(previousButton, {
          pressed: null
        });
      };

      nextButton.onmouseup = function () {
        O$.setStyleMappings(nextButton, {
          pressed: null
        });
      };

      previousButton.ondragstart =
      nextButton.ondragstart =
      previousButton.onselectstart =
      nextButton.onselectstart = function(e) {
        O$.breakEvent(e);
      };

      previousButton.ondblclick =
      nextButton.ondblclick = O$.repeatClickOnDblclick;

    }

    switcher._getTimeTable = function() {
      if (!switcher._timeTable) {
        switcher._timeTable = O$(switcher._timeTableId);
      }
      return switcher._timeTable;
    };

    switcher.getFirstDay = function() {
      return switcher._firstDay;
    };

    switcher.setFirstDay = function(firstDay) {
      if (O$._datesEqual(switcher._firstDay, firstDay))
        return;
      switcher._firstDay = firstDay;

      if (switcher._pattern) {
        var dtf = O$.getDateTimeFormatObject(switcher._locale);
        var lastDay = O$.incDay(firstDay, 6);

        var text = dtf.format(firstDay, switcher._pattern)
            .concat(switcher._splitter)
            .concat(dtf.format(lastDay, switcher._pattern));
        if (!isChrome()) {
          switcher._text.innerHTML = text;
        } else {
          var cloned = switcher._text.cloneNode();
          var origin = switcher._text;
          cloned.innerHTML = text;
          switcher._text.parentNode.replaceChild(cloned, origin);
          switcher._text = cloned;
        }
      }

      var timeTable = switcher._getTimeTable();
      if (!timeTable) {
        return;
      }

      switcher._timeTable.setDay(firstDay)
    };

    isChrome = function() {
      return Boolean(window.chrome);
    };

    switcher.previousPeriod = function() {
      var prevDay = O$.incDay(switcher._firstDay, -7);
      switcher.setFirstDay(prevDay);
    };

    switcher.nextPeriod = function() {
      var nextDay = O$.incDay(switcher._firstDay, 7);
      switcher.setFirstDay(nextDay);
    };

    if (enabled) {
      previousButton.onmousedown = function () {
        O$.setStyleMappings(previousButton, {
          pressed: stylingParams.previousButtonPressedClass
        });
        switcher.previousPeriod();
      };

      nextButton.onmousedown = function () {
        O$.setStyleMappings(nextButton, {
          pressed: stylingParams.nextButtonPressedClass
        });
        switcher.nextPeriod();
      };

      O$.addLoadEvent(function () {
        var timeTable = switcher._getTimeTable();
        if (!timeTable) {
          return;
        }

        var _onWeekChange = timeTable._onWeekChange;
        timeTable._onWeekChange = function(firstDay) {
          if (_onWeekChange) {
            _onWeekChange(firstDay);
          }
          switcher.setFirstDay(firstDay);
        };
      });

    }


  }
};

File: openFaces/source/org/openfaces/component/timetable/WeekSwitcher.java
/*
 * OpenFaces - JSF Component Library 3.0
 * Copyright (C) 2007-2012, TeamDev Ltd.
 * licensing@openfaces.org
 * Unless agreed in writing the contents of this file are subject to
 * the GNU Lesser General Public License Version 2.1 (the "LGPL" License).
 * This library is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * Please visit http://openfaces.org/licensing/ for more details.
 */

package org.openfaces.component.timetable;

import org.openfaces.util.ValueBindings;

import javax.faces.context.FacesContext;

/**
 * @author Roman Gorodischer
 */
public class WeekSwitcher extends AbstractSwitcher<WeekTable> {
    public static final String COMPONENT_TYPE = "org.openfaces.WeekSwitcher";
    public static final String COMPONENT_FAMILY = "org.openfaces.WeekSwitcher";
    private String fromPattern;
    private String toPattern;

    public WeekSwitcher() {
        setRendererType("org.openfaces.WeekSwitcherRenderer");
    }

    public String getFamily() {
        return COMPONENT_FAMILY;
    }

    @Override
    public Object saveState(FacesContext context) {
        return new Object[]{
                super.saveState(context),
                fromPattern,
                toPattern
        };
    }

    @Override
    public void restoreState(FacesContext context, Object stateObj) {
        Object[] state = (Object[]) stateObj;
        int i = 0;
        super.restoreState(context, state[i++]);
        fromPattern = (String) state[i++];
        toPattern = (String) state[i++];
    }

    @Override
    public Timetable.ViewType getApplicableViewType() {
        return Timetable.ViewType.WEEK;
    }

    public String getFromPattern() {
        String fromPattern = ValueBindings.get(this, "fromPattern", this.fromPattern);
        if (fromPattern == null) {
            TimePeriodSwitcher timePeriodSwitcher = getTimePeriodSwitcher();
            if (timePeriodSwitcher != null)
                fromPattern = timePeriodSwitcher.getFromWeekPattern();
        }
        return fromPattern;
    }

    public void setFromPattern(String fromPattern) {
        this.fromPattern = fromPattern;
    }

    public String getToPattern() {
        String toPattern = ValueBindings.get(this, "toPattern", this.toPattern);
        if (toPattern == null) {
            TimePeriodSwitcher timePeriodSwitcher = getTimePeriodSwitcher();
            if (timePeriodSwitcher != null)
                toPattern = timePeriodSwitcher.getToWeekPattern();
        }

        return toPattern;
    }

    public void setToPattern(String toPattern) {
        this.toPattern = toPattern;
    }
}
